### PR TITLE
Feature/project manipulation

### LINKED ISSUE #4: Project Manipulation
## Requested Functionality:

Follow a TDD approach

Add the repository and its implementation needed for this task.

### Main functions:

- Add project
- Remove project
- Edit project
- Assign Mate

### PR DESCRIPTION
**Pull request by Mohamed-Boumaraf**
## Add Project Repository Interface with implementation and unit tests

This PR introduces the implementation of the Project Manipulation feature, including:

-  **Interface**: `ProjectRepository` defines the interface for project persistence and mate assignment operations.
    
-  **Implementation**: `ProjectRepositoryImpl` backed by a generic `DataSource<Project>`, with robust logic for:
    - Creating, updating, deleting, and retrieving projects
    - Assigning and unassigning mates to projects
    - Existence checks and error handling with `ProjectNotFoundException`
        
-  **Unit Tests**: Thorough test suite for all repository methods using MockK and Truth, including :
    - Test coverage includes edge cases and ensures data consistency
    - **Test-Driven Development (TDD)** approach: all functionality was written based on previously defined failing tests
    closes #4

### CODE REVIEW

File: src/main/kotlin/data/repo/ProjectRepositoryImpl.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+import logic.model.Project
+import org.damascus.data.DataSource
+import org.damascus.logic.repository.ProjectRepository
+import java.util.*
+
+
+class ProjectNotFoundException(projectId: UUID) :
+    RuntimeException("Project with ID $projectId not found")
```
Comment (yusufnasserdev) [inline src/main/kotlin/data/repo/ProjectRepositoryImpl.kt]: Include this in a separate file.

File: src/main/kotlin/data/repo/ProjectRepositoryImpl.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        val existingProjects = projectDataSource.read()
+
+        val project = existingProjects.firstOrNull { it.id == projectId }
+            ?: return false
+
+        if (project.assignedMatesIds.contains(mateId))
+            return false
+
+        val updatedProject = project.copy(
+            assignedMatesIds = project.assignedMatesIds.toMutableList().apply { add(mateId) }
```
Comment (yusufnasserdev) [inline src/main/kotlin/data/repo/ProjectRepositoryImpl.kt]: No need to cast the list, it's already a `MutableList`

File: src/main/kotlin/data/repo/ProjectRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        if (!project.assignedMatesIds.contains(mateId))
+            return false
+
+        val updatedProject = project.copy(
+            assignedMatesIds = project.assignedMatesIds.toMutableList().apply { remove(mateId) }
+        )
+
+        projectDataSource.update(project.id, updatedProject)
+        return true
+    }
```
Comment (yusufnasserdev) [inline src/main/kotlin/data/repo/ProjectRepositoryImpl.kt]: - These two functions logic is similar and it violate the DRY principle, should modularize the mutual part and use it as a function.

- Also, they represent a use-case, so they should be separated from the repo and use it as a dependency.

File: src/main/kotlin/data/repo/ProjectRepositoryImpl.kt
Comment (yusufnasserdev) [inline src/main/kotlin/data/repo/ProjectRepositoryImpl.kt:1]: - The use of `val existingProjects = projectDataSource.read()` is inconsistent throughout the file.

- Should add a function that retrieves projects assigned for a certain mate `getAllProjectsByMateId(mateId: UUID)`.

File: src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    lateinit var repo: ProjectRepositoryImpl
+
+    @BeforeEach
+    fun setup() {
+        dataSource = mockk(relaxed = true)
+        repo = ProjectRepositoryImpl(dataSource)
+    }
+
+    @Test
+    fun `should return false when create an existing project`() {
```
Comment (yusufnasserdev) [inline src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt]: Please specify what function are you testing before `should`

File: src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    @Test
+    fun `should throw ProjectNotFoundException when project does not exist`() {
+        // Given
+        val nonExistentId = UUID.randomUUID()
+        every { dataSource.read() } returns listOf()
+
+        // When
+        val exception = org.junit.jupiter.api.assertThrows<ProjectNotFoundException> {
+            repo.get(nonExistentId)
+        }
```
Comment (yusufnasserdev) [inline src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt]: Import the needed function instead of using its entire path.

File: src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+        // Then
+        assertThat(exception.message).contains(nonExistentId.toString())
```
Comment (yusufnasserdev) [inline src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt]: The `When` should be `When & Then`, please avoid using multiple assertions in a single test.

File: src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt
Comment (yusufnasserdev) [inline src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt:1]: Please reconsider the usage of `JustRun` if it's meant to verify a certain call with certain parameters.

**Review by yusufnasserdev — CHANGES_REQUESTED** [review]
Please resolve the requested changes.

Great work, keep it up!

File: src/main/kotlin/data/repo/ProjectRepositoryImpl.kt [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import logic.model.Project
+import org.damascus.data.DataSource
+import org.damascus.logic.repository.ProjectRepository
+import java.util.*
+
+
+class ProjectNotFoundException(projectId: UUID) :
+    RuntimeException("Project with ID $projectId not found")
+
+class ProjectRepositoryImpl(private val projectDataSource: DataSource<Project>) : ProjectRepository {
```
Comment (Al-Taie) [inline src/main/kotlin/data/repo/ProjectRepositoryImpl.kt]: I prefer to use `dataSource` instead of `projectDataSource`.

File: src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        assertThat(result).isTrue()
+    }
+
+}
+
+fun makeMate(mateID: UUID): Mate {
+    return Mate(mateID, "name", "password", Role.MATE)
+}
+
+fun makeProject(projectID: UUID): Project {
```
Comment (Al-Taie) [inline src/test/kotlin/data/repo/ProjectRepositoryImplTest.kt]: `makeMate` has one parameter so it's fine to name it `id` instead of `mateID` (same for `makeProject`).

**Review by yusufnasserdev — APPROVED** [review]
